src/core/mesh/qgsmeshadvancedediting.h

```diff
@@ -0,0 +1,122 @@
+/***************************************************************************
+  qgsmeshadvancedediting.h - QgsMeshAdvancedEditing
+
+ ---------------------
+ begin                : 9.7.2021
+ copyright            : (C) 2021 by Vincent Cloarec
+ email                : vcloarec at gmail dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+#ifndef QGSMESHADVANCEDEDITING_H
+#define QGSMESHADVANCEDEDITING_H
+
+
+#include "qgis_core.h"
+#include "qgstopologicalmesh.h"
+#include "qgstriangularmesh.h"
+
+class QgsMeshEditor;
+class QgsProcessingFeedback;
+
+/**
+ * \ingroup core
+ *
+ * \brief Abstract class that can be derived to implement advanced editing on mesh
+ *
+ * To apply the advanced editing, a pointer to and instance of a derived class is passed
+ * in the method QgsMeshEditor::advancedEdit().
+ *
+ * \since QGIS 3.22
+ */
+class CORE_EXPORT QgsMeshAdvancedEditing : protected QgsTopologicalMesh::Changes SIP_ABSTRACT
+{
+  public:
+
+    //! Constructor
+    QgsMeshAdvancedEditing();
+    //! Destructor
+    virtual ~QgsMeshAdvancedEditing();
+
+    //! Sets the input vertices indexes that will be used for the editing
+    void setInputVertices( const QList<int> verticesIndexes );
+
+    //! Sets the input faces indexes that will be used for the editing
+    void setInputFaces( const QList<int> faceIndexes );
+
+    //! Returns a message provided by the advanced editing after
+    QString message() const;
+
+    //! Removes all data provided to the editing or created by the editing
+    void clear();
+
+  protected:
+    QList<int> mInputVertices;
+    QList<int> mInputFaces;
+    QString mMessage;
+
+    /**
+     * Apply a change to \a mesh Editor. This method is called by the QgsMeshEditor to apply the editing on the topological mesh
+     *
+     * The method has to be implemented in the derived class to provide the changes of the advancd editing
+     */
+    virtual QgsTopologicalMesh::Changes apply( QgsMeshEditor *meshEditor ) = 0; SIP_SKIP
+
+    friend class QgsMeshEditor;
+};
+
+/**
+ * \ingroup core
+ *
+ * \brief Class that can do a refinement of faces of a mesh
+ *
```

Comment:
can you please add description what the face refinement is and how it is algorithmically performed? also that it works only on faces with 3 or 4 vertices?